src/legacy/ui/public/vis/vis_types/vislib_vis_legend/vislib_vis_legend.tsx

```diff
@@ -0,0 +1,264 @@
+        ],
+      });
+      return;
+    } // make sure vislib is defined at this point
+
+    if (
+      this.props.uiState.get('vis.legendOpen') == null &&
+      this.props.vis.params.addLegend != null
+    ) {
+      this.setState({ open: this.props.vis.params.addLegend });
+    }
+
+    if (CUSTOM_LEGEND_VIS_TYPES.includes(vislibVis.visConfigArgs.type)) {
+      const legendLabels = this.props.vislibVis.getLegendLabels();
+      if (legendLabels) {
+        this.setState({
+          labels: map(legendLabels, label => {
+            return { label };
+          }),
+        });
+      }
+    } else {
+      this.setState({ labels: this.getLabels(this.props.visData, vislibVis.visConfigArgs.type) });
+    }
+
+    if (vislibVis.visConfig) {
+      this.getColor = this.props.vislibVis.visConfig.data.getColorFunc();
+    }
+
+    this.setState({ tableAggs: getTableAggs(this.props.vis) });
+  };
+
+  highlight = (event: BaseSyntheticEvent) => {
+    const el = event.currentTarget;
+    const handler = this.props.vislibVis && this.props.vislibVis.handler;
+
+    // there is no guarantee that a Chart will set the highlight-function on its handler
+    if (!handler || typeof handler.highlight !== 'function') {
+      return;
+    }
+    handler.highlight.call(el, handler.el);
+  };
+
+  unhighlight = (event: BaseSyntheticEvent) => {
+    const el = event.currentTarget;
+    const handler = this.props.vislibVis && this.props.vislibVis.handler;
+
+    // there is no guarantee that a Chart will set the unhighlight-function on its handler
+    if (!handler || typeof handler.unHighlight !== 'function') {
+      return;
+    }
+    handler.unHighlight.call(el, handler.el);
+  };
+
+  getAnchorPosition = () => {
+    const { position } = this.props;
+
+    switch (position) {
+      case 'bottom':
+        return 'upCenter';
+      case 'left':
+        return 'rightUp';
+      case 'right':
+        return 'leftUp';
+      default:
+        return 'downCenter';
+    }
+  };
+
+  renderLegend = (anchorPosition: EuiPopoverProps['anchorPosition']) => (
+    <ul className="visLegend__list" id={this.legendId}>
+      {this.state.labels.map(item => (
+        <VisLegendItem
+          item={item}
+          key={item.label}
+          anchorPosition={anchorPosition}
+          selected={this.state.selectedLabel === item.label}
+          canFilter={this.canFilter(item)}
+          onFilter={this.filter}
+          onSelect={this.toggleDetails}
+          legendId={this.legendId}
+          setColor={this.setColor}
+          getColor={this.getColor}
+          onHighlight={this.highlight}
+          onUnhighlight={this.unhighlight}
+        />
+      ))}
+    </ul>
+  );
+
+  render() {
+    const { open } = this.state;
+    const anchorPosition = this.getAnchorPosition();
+
+    return (
+      <div className="visLegend">
+        <button
+          type="button"
+          onClick={this.toggleLegend}
+          className={classNames('visLegend__toggle', {
```

Comment:
Just one extra thing to match that of the overflow menu button. Can you add this class
```suggestion
          className={classNames('visLegend__toggle kbn-resetFocusState', {
```

Reply:
67d3172